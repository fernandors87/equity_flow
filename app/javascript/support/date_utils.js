import { List } from 'immutable'
import moment from 'moment'

export function range(startDate, endDate, period) {
  if (startDate.isAfter(endDate)) {
    throw `startDate(${startDate}) must not be greater than endDate(${endDate})`
  }

  const startOfPeriod = moment(startDate).startOf(period)
  const endOfPeriod = moment(endDate).endOf(period)
  const generateRange = (acc) => {
    const lastProcessed = acc.last()
    const plusOne = moment(lastProcessed).add(1, period)
    return plusOne.isBefore(endOfPeriod) ? generateRange(acc.push(plusOne)) : acc
  }
  return generateRange(List.of(startOfPeriod))
}

export function utc(date) {
  return moment.utc(date).utcOffset(0, true)
}
