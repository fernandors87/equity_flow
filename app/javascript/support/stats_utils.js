export function sum(series) {
  return series.reduce((sum, x) => {
    if (typeof x !== 'number') throw `'${x}' is not a number and cannot be summed`
    return sum + x
  }, 0)
}

export function mean(series) {
  if (series.isEmpty()) return 0
  return sum(series) / series.size
}
