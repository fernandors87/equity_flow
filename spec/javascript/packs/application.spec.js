import Application from 'packs/application'
import {Set} from 'immutable'
import {splitFix} from '../support/fixtures'
import moment from 'moment'

describe('Application', () => {

  const api = {
    splits: {
      list: sinon.stub().returns(Promise.resolve(Set([splitFix]))),
    },
    accounts: {
      list: sinon.stub().returns(Promise.resolve(Set()))
    }
  }

  it('render the component', () => {
    const wrapper = shallow(<Application api={api}/>)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('update splits according to date range', () => {
    const startDate = moment.utc('2017-09-11T00:00:00.000+00:00')
    const endDate = moment.utc('2018-01-07T00:00:00.000+00:00')

    const wrapper = mount(<Application api={api} />)
    wrapper.instance()._onDateChange(startDate, endDate)
    expect(api.splits.list.callCount).toEqual(3)
    process.nextTick(() => {
      const {startDate, endDate, splits} = wrapper.state()
      expect(splits).toEqualImmutable(Set([splitFix]))
      expect(startDate).toEqualMoment(startDate)
      expect(endDate).toEqualMoment(endDate)
    })
  })
})
