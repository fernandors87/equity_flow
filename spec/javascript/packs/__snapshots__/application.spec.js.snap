// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application render the component 1`] = `"<div><div class=\\"datepicker form-group has-feedback\\"><div><input type=\\"text\\" readonly=\\"\\" container=\\"[object Object]\\" value=\\"July 1, 2017 - February 8, 2018\\" class=\\"datepicker-input form-control\\"/><span class=\\"form-control-feedback glyphicon glyphicon-calendar\\"></span></div></div><div><table class=\\"table\\"><thead><tr><th>Account</th><th>2017-07</th><th>2017-08</th><th>2017-09</th><th>2017-10</th><th>2017-11</th><th>2017-12</th><th>2018-01</th><th>2018-02</th><th>Mean</th></tr></thead><tbody></tbody></table></div></div>"`;
